### PR TITLE
sync: retry on failure

### LINKED ISSUE #1468: [Background sync] Report failures and retry

### PR DESCRIPTION
**Pull request by scolsen**
Updates the survey sync worker and tileset download worker to retry sync
on certain failures. In addition, adds unit tests to verify the workers
return the expected results.

Fixes #1468

### CODE REVIEW

File: ground/src/main/java/com/google/android/ground/persistence/sync/SurveySyncWorker.kt [hunk 1]
```diff
@@ -44,20 +44,23 @@ constructor(
...
-      syncSurvey(surveyId)
+    try {
+      // It's ok to block here since WorkManager calls doWork() on a background thread.
+      runBlocking {
+        Timber.d("Syncing survey $surveyId")
+        syncSurvey(surveyId)
+      }
+    } catch (e: Throwable) {
+      Timber.e("error syncing survey in the background", e)
+      return Result.retry()
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/persistence/sync/SurveySyncWorker.kt:55]: How many times will this retry the work request?

File: ground/src/test/java/com/google/android/ground/persistence/sync/SurveySyncWorkerTest.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          inputData = workDataOf(Pair(SURVEY_ID_PARAM_KEY, SURVEY.id)),
+        )
+        .setWorkerFactory(factory)
+        .build()
+    val result = worker.doWork()
+    assertThat(result).isEqualTo(Result.success())
+  }
+
+  @Test
+  fun doWork_RetriesOnFailure() = runWithTestDispatcher {
```
Comment (shobhitagarwal1612) [inline ground/src/test/java/com/google/android/ground/persistence/sync/SurveySyncWorkerTest.kt:100]: Will this retry the failed work indefinitely?

File: ground/src/test/java/com/google/android/ground/persistence/sync/SurveySyncWorkerTest.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  fun doWork_FailsOnNoInput() {
+    val worker =
+      TestWorkerBuilder<SurveySyncWorker>(
+          context,
+          executor,
+          inputData = workDataOf(Pair(SURVEY_ID_PARAM_KEY, null)),
+        )
+        .setWorkerFactory(factory)
+        .build()
+    val result = worker.doWork()
```
Comment (shobhitagarwal1612) [inline ground/src/test/java/com/google/android/ground/persistence/sync/SurveySyncWorkerTest.kt:79]: Can we create an abstract class for creating the worker along with the dependencies and calling `.doWork()`? 

Something like:

```
fun createWorkerAndDoWork<T>(args: Pair) {}
```

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: ground/src/test/java/com/google/android/ground/persistence/sync/SurveySyncWorkerTest.kt (hunk 3 above)
Comment (scolsen) [reply ground/src/test/java/com/google/android/ground/persistence/sync/SurveySyncWorkerTest.kt:79]: If we wind up needing more of these that sounds like a good idea; for now I don't think the extra indirection is worth it yet.

File: ground/src/main/java/com/google/android/ground/persistence/sync/SurveySyncWorker.kt (hunk 1 above)
Comment (shobhitagarwal1612) [reply ground/src/main/java/com/google/android/ground/persistence/sync/SurveySyncWorker.kt:59]: Thanks

**Review by shobhitagarwal1612 — APPROVED** [review]

**Pull request merged**